[Documentation] Defaulting guidelines (#9122)
gardener/gardener: docs/development/defaulting.md

> # Defaulting Strategy and Developer Guideline

Reviewer: You could add a link to the docs in the index (`docs/README.md`).

This document walks you through:

> - Conventioned to be followed when writing defaulting functions

Reviewer: I guess you rather mean

```suggestion
- Conventions to be followed when writing defaulting functions
```

?

- How to write test for defaulting function

The document is aimed towards developers that want to contribute code and need to write defaulting code and unit tests covering the defaulting functions, as well as maintainers and reviewers that review code.
It serves as a common guide that we commit to follow in our project to ensure consistency in our defaulting code, good coverage for high confidence, and good maintainability.

## Writing defaulting code

> - Every Kubernetes type should have a dedicated `defaults_*.go` file. For instance, if you have a `Shoot` type, there should be a corresponding `defaults_shoot.go` file containing all defaulting logic for that type.

Reviewer: You could link example from the code.

> - If there is only one type under an api group then we can just have `types.go` and a corresponding `defaults.go` file.

Reviewer: You could link example from the code (if we have any).

> - Aim to segregate each struct type into its own `SetDefaults_*` function. These functions encapsulate the defaulting logic specific to the corresponding struct type, enhancing modularity and maintainability.

Reviewer: You could link example from the code.

> :warning: Ensure to run the `make generate` command, which generates the `zz_generated.defaults.go` file containing the overall defaulting function.

Reviewer: I guess you rather mean that `make generate` has to be executed after a new `SetDefaults_*` func is added? Otherwise, reading this sentence I don't understand when I have to execute `make generate`.

Reviewer: Additionally, there should be a more effective way of generation, for example executing `make generate WHAT=codegen` instead of `make generate`.

## Writing unit tests for defaulting code

> - Each test case should validate the overall defaulting function `SetObjectDefaults_*` generated by `defaulter-gen` and not a specific `SetDefaults_*`. This way we also test if the `zz_generated.defaults.go` was generated correctly.

Reviewer: You could link example from the code.

- Test each defaulting function carefully to ensure:
    - Proper defaulting behaviour when fields are empty or nil. Note that some fields may be optional and should not be defaulted.

>     - Preservation of existing values, ensuring that defaulting does not accidentally overwrite them.

Reviewer: You could link example from the code.

# Defaulting Strategy and Developer Guideline

This document walks you through:

- Conventioned to be followed when writing defaulting functions
- How to write test for defaulting function

The document is aimed towards developers that want to contribute code and need to write defaulting code and unit tests covering the defaulting functions, as well as maintainers and reviewers that review code.
It serves as a common guide that we commit to follow in our project to ensure consistency in our defaulting code, good coverage for high confidence, and good maintainability.

The guidelines are not meant to be absolute rules.
Always apply common sense and adapt the guideline if it doesn't make much sense for some cases.
If in doubt, don't hesitate to ask questions during a PR review (as an author, but also as a reviewer).

> Add new learnings as soon as we make them!

Reviewer 2: Sorry but I don't understand what you meant by sentence on line 14

Author: These lines are removed now.

## Writing defaulting code

- Every Kubernetes type should have a dedicated `defaults_*.go` file. For instance, if you have a `Shoot` type, there should be a corresponding `defaults_shoot.go` file containing all defaulting logic for that type.

> - If the there is only type under an api group then we can have `types.go` and corresponding `defaults.go` file.

Reviewer 2: ```suggestion
- If there is only one type under an api group then we can just have `types.go` and a corresponding `defaults.go` file.
```

- Aim to segregate each struct type into its own `SetDefaults_*` function. These functions encapsulate the defaulting logic specific to the corresponding struct type, enhancing modularity and maintainability.

:warning: Ensure to run the `make generate` command, which generates the `zz_generated.defaults.go` file containing the overall defaulting function.

## Writing unit tests for defaulting code

> - Each test case should validate the overall defaulting function `SetObjectDefaults_*` generated by `defaulter-gen`.

Reviewer 2: ```suggestion
- Each test case should validate the overall defaulting function `SetObjectDefaults_*` generated by `defaulter-gen` and not a specific `SetDefaults_*`. This way we also test if the `zz_generated.defaults.go` was generated correctly.
```

- Test each defaulting function carefully to ensure:

>     - Proper defaulting behavior when fields are empty or nil.

Reviewer 2: ```suggestion
    - Proper defaulting behaviour when fields are empty or nil. Note that some fields may be optional and should not be defaulted.
```

> :warning: Ensure to run the `make generate WHAT=codegen` command whenever any changes are done in defaulting logic or a new `SetDefaults_*` function is added, which generates the `zz_generated.defaults.go` file containing the overall defaulting function.

Reviewer: I am not sure about ` whenever any changes are done in defaulting logic`? What do you mean with it? It is enough to run the generation only when a `SetDefaults_*` function is added or removed?

> - Each test case should validate the overall defaulting function `SetObjectDefaults_*` generated by `defaulter-gen` and not a specific `SetDefaults_*`. This way we also test if the `zz_generated.defaults.go` was generated correctly. For example, `cloudprofile` defaulting should be tested with overall defaulting function `SetObjectDefaults_CloudProfile()`.
> https://github.com/gardener/gardener/blob/ff5a5be6049777b0695659a50189e461e1b17796/pkg/apis/core/v1beta1/defaults_cloudprofile_test.go#L40-L47

Reviewer: ```suggestion
- Each test case should validate the overall defaulting function `SetObjectDefaults_*` generated by `defaulter-gen` and not a specific `SetDefaults_*`. This way we also test if the `zz_generated.defaults.go` was generated correctly.
    For example, the `spec.machineImages[].updateStrategy` field in the CloudProfile is defaulted as follows:
    https://github.com/gardener/gardener/blob/ff5a5be6049777b0695659a50189e461e1b17796/pkg/apis/core/v1beta1/defaults_cloudprofile.go#L23-L29
    The defaulting should be tested with the overall defaulting function `SetObjectDefaults_CloudProfile` (and not with `SetDefaults_MachineImage`):
    https://github.com/gardener/gardener/blob/ff5a5be6049777b0695659a50189e461e1b17796/pkg/apis/core/v1beta1/defaults_cloudprofile_test.go#L40-L47
```

>         For example, when `spec.SecretRef.Namespace` field of `SecretBinding` is nil, it should be defaulted to the namespace of SecretBinding object. But `spec.SecretRef.Namespace` field should not be overwritten by defaulting logic if it is already set.

Reviewer: ```suggestion
        For example, when `spec.secretRef.namespace` field of `SecretBinding` is nil, it should be defaulted to the namespace of SecretBinding object. But `spec.secretRef.namespace` field should not be overwritten by defaulting logic if it is already set.
```

> # Defaulting Strategy and Developer Guideline

Reviewer 3: ```suggestion
# Defaulting Strategy and Developer Guidelines
```

This document walks you through:

- Conventions to be followed when writing defaulting functions

> - How to write test for defaulting function

Reviewer 3: ```suggestion
- How to write a test for a defaulting function
```

> The document is aimed towards developers that want to contribute code and need to write defaulting code and unit tests covering the defaulting functions, as well as maintainers and reviewers that review code.

Reviewer 3: ```suggestion
The document is aimed towards developers who want to contribute code and need to write defaulting code and unit tests covering the defaulting functions, as well as maintainers and reviewers who review code.
```
